test_semver.py

```diff
@@ -497,3 +497,9 @@ def test_immutable_unknown_attribute(version):
 def test_version_info_should_be_iterable(version):
     assert tuple(version) == (version.major, version.minor, version.patch,
                               version.prerelease, version.build)
+
+
+def test_should_compare_prerelease_and_build_with_numbers():
+    v1 = VersionInfo(major=1, minor=9, patch=1, prerelease=1, build=1)
+    v2 = VersionInfo(major=1, minor=9, patch=1, prerelease=2, build=1)
```

Comment:
Minor: I suggest to add a test with ` VersionInfo(3, 2, 1, 1)` style ctor as used in the commit description.

Reply:
@Flamefire Thanks for the suggestion! :+1:  I've added this in the next commits (see below).

Reply:
> Doesn't `**v1` expand the dictionary to kwargs? If so then there is no `VersionInfo(3, 2, 1, 1)` usage. 

Well, yes, you could replace the dictionary with a tuple. Would be the same. :wink: 


> PS: I would not use the `parametrize` feature here. It isn't really shorter and obscures the meaning more than it helps (see my confusion above). Why not put a couple asserts in the tests instead? 

I think you are right, the test could be simplified.

> This would also double as documentation:
> 
> ```
> assert VersionInfo(major=1, minor=9, patch=1, prerelease=1, build=1) <  VersionInfo(major=1, minor=9, patch=1, prerelease=2, build=1)
> assert VersionInfo(1, 9, 1, 1, 1) <  VersionInfo(1, 9, 1, 2, 1)
> assert VersionInfo(10) <  VersionInfo('2')
> assert VersionInfo('10') <  VersionInfo('2')
> ```

I think, the last two conditions are wrong. At least, the test function gives me an AssertionError. I think what you mean is this:

```python
assert VersionInfo('2') <  VersionInfo(10)
assert VersionInfo('2') <  VersionInfo('10')
```

Correct? Can you confirm?

Reply:
Yes correct of course, C&P ;-) I also mistakenly added a double space, so make sure to remove it on the patch

Reply:
Haha, ok. No problem. I will add it and you as a co-author. Thanks! :+1:

Reply:
@Flamefire  Hmn, after looking at the function name of the test case, it seems it doesn't really show its intention. Unfortunately, I didn't have a better idea.

Maybe do you have a better name? :wink:  Or any other things I should change?

Reply:
Maybe: Split the 2nd part into a new test like `test_should_be_able_to_use_strings_as_major_minor_patch` and add additional asserts for the minor and patch. Maybe it is even enough to do:

```
v = VersionInfo('1', '2', '3')
assert v.major == 1
assert isinstance(v.major, int)
...
assert v.prerelease is None
assert v.build is None
```

And a similar test for the int->string in prerelease&build

Reply:
Good idea! :+1: I have one question regarding this code:

```python
assert v.major == 1
assert isinstance(v.major, int)
```

Could we avoid the first assert? The reason why I'm asking is: this is covered already in other tests. IMHO, we don't need to repeat ourselves. In this test we are interested in the correct *type* only, not the correct *content*, right?

If you agree, it would simplify the test a bit and focus on the type check with `isinstance`. With that, I end up with this test:

```python
def test_should_be_able_to_use_strings_as_major_minor_patch():
    v = VersionInfo('1', '2', '3')
    assert isinstance(v.major, int)
    assert isinstance(v.minor, int)
    assert isinstance(v.patch, int)
    assert v.prerelease is None
    assert v.build is None
```

Ok with that? :-)

Reply:
I would add an assertion `VersionInfo('1', '2', '3') == VersionInfo(1, 2, 3)` so a typo like ` self._minor = int(major)` is caught. While I agree that this tests the *type* it must also assure that the conversion is right. Maybe someone implements it as `int(major) if isintance(major, string) else major` later. This is not covered by any other test. The above assertion is also very good for documentation as it is as descriptive as it gets.

Reply:
Yes, you're right. I've added your additional line and my above lines in commit a611a73. Better? :wink:

Reply:
Added a comment  on the line with suggestion :)